include/cutlass/gemm/kernel/intel_pvc_tile_scheduler_streamk.hpp

```diff
@@ -0,0 +1,692 @@
+    uint32_t num_accumulator_mtxs = 1) {
+
+    using ElementAccumulator = typename FrgTensorC::value_type;
+
+    if (!requires_fixup(params, work_tile_info)) {
+      return;
+    }
+    auto tile_idx = output_tile_index(params, work_tile_info);
+
+    // Index of the lock on which to wait
+    auto lock_idx = (tile_idx * num_barriers) + barrier_idx;
+
+    auto reduction_tile_idx = tile_idx;
+    auto reduction_peer_offset = 0;
+    int barrier_group_thread_idx = ThreadIdxX();
+
+    // Reductions use BlockStripedReduce with a width of BarrierManager::ThreadCount under the hood.
+    // Thus, the start of the reduction space is the same across all threads in a warp group.
+    int reduction_offset =
+      (cute::size<0>(TileShape{}) * cute::size<1>(TileShape{}) * reduction_tile_idx * num_accumulator_mtxs) +
+      reduction_peer_offset;
+
+    ElementAccumulator* group_reduction_workspace = reinterpret_cast<ElementAccumulator*>(params.reduction_workspace_) + reduction_offset;
+
+    using AccumulatorArrayT = Array<typename FrgTensorC::value_type, size(FrgTensorC{})>;
+    using BlockStripedReduceT = BlockStripedReduce<ThreadsPerBlock, AccumulatorArrayT>;
+
+    AccumulatorArrayT* reduction_workspace_array = reinterpret_cast<AccumulatorArrayT*>(group_reduction_workspace);
+    AccumulatorArrayT* accumulator_array = reinterpret_cast<AccumulatorArrayT*>(accumulators.data());
+
+    // The number of tiles for which reduction is required is either:
+    //   (a) the total number of output tiles (in the case of split-K)
+    //   (b) the number of stream-K tiles
+    // To calculate the total number of output tiles in the split-K case, we
+    // note that, in the split-K case, the units_per_problem_ member of Params will be
+    // the total number of output tiles.
+    uint32_t reduction_tiles = 0;
+    if (params.divmod_splits_.divisor > 1) {
+      reduction_tiles = params.units_per_problem_;
+    }
+    else {
+      reduction_tiles = params.sk_tiles_;
+    }
+
+    auto reduction_workspace_size = Params::get_reduction_workspace_size(
+      reduction_tiles, to_gemm_coord(TileShape{}), sizeof_bits<ElementAccumulator>::value, num_accumulator_mtxs);
+    BarrierType* lock_workspace = reinterpret_cast<BarrierType*>(
+      reinterpret_cast<uint8_t*>(params.reduction_workspace_) + reduction_workspace_size);
+
+    if (!compute_epilogue(work_tile_info, params)) {
+      if (work_tile_info.K_idx == 0) {
+        // The first peer initializes the workspace partials
+        BlockStripedReduceT::store(reduction_workspace_array, *accumulator_array, barrier_group_thread_idx);
+      }
+      else {
+        // Wait until the preceding split added its accumulators
+        BarrierManager::wait_eq(barrier_idx, lock_workspace, barrier_group_thread_idx, lock_idx, work_tile_info.K_idx);
+
+        // Perform reduction in workspace
+        BlockStripedReduceT::reduce(reduction_workspace_array, *accumulator_array, barrier_group_thread_idx);
```

Comment:
From what I understand, We are using the same `BlockStripedReduceT` that nvidia uses (unless I probably missed it). The definition of it is in `block_striped.h`, and by the looks of it, it is very nvidia centric, in the sense that it revolves around the concept of coalesced loads/stores, and reinterpret casts into the array type, such that it leads to the vector loads in the PTX instructions...

Should we be using 1D load / store intrinsics (maybe even 2D, if we want one workgroup to reduce more than one K, but that might require extra synchronization maybe).
Would the `reinterpret_cast`s in `block_striped.h`  lead to to a similar effect ?

Reply:
Just one last question on this,

did we consider partial shuffle reduction at the sub-group level and then the warp leader (or half warp, depending on how many shuffles we want to do) atomically adds to the global memory.  That might lead to lesser global memory atomic operations ?